go/libraries/doltcore/sqle/setalgebra/interval.go

```diff
@@ -0,0 +1,293 @@
+	// interval it will also return true if it is equal to the start value
+	var nbf *types.NomsBinFormat
+	matchesStartCondition := func() (bool, error) {
+		res, err := in.Start.Val.Less(nbf, val)
+
+		if err != nil {
+			return false, err
+		}
+
+		if res {
+			return res, nil
+		}
+
+		if in.Start.Inclusive {
+			return in.Start.Val.Equals(val), nil
+		}
+
+		return false, nil
+	}
+
+	// matchesStartCondition returns true if the value is less than the start.  For a closed
+	// interval it will also return true if it is equal to the start value
+	matchesEndCondition := func() (bool, error) {
+		res, err := val.Less(nbf, in.End.Val)
+
+		if err != nil {
+			return false, err
+		}
+
+		if res {
+			return res, nil
+		}
+
+		if in.End.Inclusive {
+			return in.End.Val.Equals(val), nil
+		}
+
+		return false, nil
+	}
+
+	// if the interval is finite (has a start and end value) the result is true if both the start and
+	// end condition check return true
+	if in.Start != nil && in.End != nil {
+		st, err := matchesStartCondition()
+
+		if err != nil {
+			return false, err
+		}
+
+		end, err := matchesEndCondition()
+
+		if err != nil {
+			return false, err
+		}
+
+		return st && end, nil
+	} else if in.End == nil {
+		// No end means the interval goes to positive infinity.  All values that match the start
+		// condition are in the interval
+		return matchesStartCondition()
+	} else {
+		// No start means the interval goes to negative infinity.  All values that match the end
+		// condition are in the interval
+		return matchesEndCondition()
+	}
+}
+
+// intervalComparison contains the results of compareIntervals.
+type intervalComparison [4]int
+
+// intervalComparisonIndex is an enum which allows you to access the specific comparison of two
+// points compared by a call to compareIntervals(...)
+type intervalComparisonIndex int
+
+const (
+	start1start2 intervalComparisonIndex = iota
+	start1end2
+	end1start2
+	end1end2
+)
+
+// noOverlapLess is the result you will get when comparing 2 intervals (A,B] and [C,D) where
+// B is less than C
+var noOverlapLess = intervalComparison{-1, -1, -1, -1}
+
+// noOverlapGreater is the result you will get when comparing 2 intervals [A,B) and (C,D] where
+// A is greater than D
+var noOverlapGreater = intervalComparison{1, 1, 1, 1}
+
+// compareIntervals compares the start and end points of one interval against the start and end
+// points of another interval. The resulting intervalComparison is an array of 4 ints where a
+// value of -1 means that the point in interval 1 is less than the point in interval 2. A value
+// of 0 indicates equality, and a value of 1 indicates the value in interval 1 is greater than
+// the value in interval 2.
+func compareIntervals(in1, in2 Interval) (intervalComparison, error) {
+	comp := intervalComparison{}
+	endpoints := [2][2]*IntervalEndpoint{{in1.Start, in1.End}, {in2.Start, in2.End}}
+
+	var err error
+	for i := 0; i < 2; i++ {
```

Comment:
You only have N=4 here. This might be easier to read and verify if you just wrote 4 consecutive chunks of comparison logic, rather than structure it as a loop.

Reply:
updated